src/lib/expression/value_expression.hpp

```diff
@@ -18,6 +19,7 @@ class ValueExpression : public AbstractExpression {
   DataType data_type() const override;
 
   const AllTypeVariant value;
+  const std::optional<ValueExpressionID> value_expression_id;
```

Comment:
Why doesn't the value_expression_id affect the equality of two expressions?

Reply:
there exist value expresions with and without id since the id is only attributed to expressions that originate directly from the sql translator. however, during query processing intermediate value expressions are generated at a lot of places, where an id doesn't really make sense.
would something like ignoring the id if not both expressions have one and only comparing them if it is set for both be a solution here?

Reply:
I'm not even sure yet if we need the expression ID to be part of the equality. Right now, it either does not make a difference or our tests are not catching it.

Let us take the query `SELECT ... WHERE a > 3 AND a < 3 + 1`. Traditionally, both `ValueExpression{3}` would be considered equal. If the ExpressionReductionRule would not be marked as a rule that prevents the LQP to be reused with other parameters, we might end up in a situation where the condition `WHERE a > 3 AND a < 4 + 1` would be incorrectly reusing an existing result, correct?

I am somewhat worried about the prevents_caching flag being the only thing that keeps an LQP from becoming incorrect when executed with different literals. Right now, we do not have a safeguard against it, do we?

Reply:
> ight now, it either does not make a difference or our tests are not catching it.
we don't catch yet. 
also in most tests, the expected plans/expressions are created not via a sql query and translation but right out of functional expressions. so they dont have an id.

in my understanding, the prevents_caching flag is exactly this safeguard. i should definitely add tests that make sure that it does its job, but what other safeguard did you had in mind?

Reply:
I meant "safeguard" in the sense of "safety net". How can we identify problems that come up when someone erroneously marks their new rule as "cacheable"? I was hoping that we can find some assertions for that, but I can't come up with any right now.